src/main/java/com/redhat/devtools/lsp4ij/operations/codeactions/LSPLazyCodeActions.java

```diff
@@ -129,59 +102,57 @@ private List<Either<Command, CodeAction>> getOrLoadCodeActions() {
-                                // Language server is initialized here
-                                cancelChecker.checkCanceled();
-
-                                // Collect code action for the given file by using the language server
-                                CodeActionParams params = createCodeActionParams(diagnostic, file);
-                                return ls.getTextDocumentService()
-                                        .codeAction(params)
-                                        .thenApply(codeActions -> {
-                                            // Code action are collected here
-                                            cancelChecker.checkCanceled();
-                                            if (codeActions == null) {
-                                                return Collections.emptyList();
-                                            }
-                                            return codeActions;
-                                        });
-                            });
-                });
+                .computeAsyncCompose(cancelChecker -> languageServerWrapper
+                        .getInitializedServer()
+                        .thenCompose(ls -> {
+                            // Language server is initialized here
+                            cancelChecker.checkCanceled();
+
+                            // Collect code action for the given file by using the language server
+                            CodeActionParams params = createCodeActionParams(diagnostics, file);
+                            return ls.getTextDocumentService()
+                                    .codeAction(params)
+                                    .thenApply(codeActions -> {
+                                        // Code action are collected here
+                                        cancelChecker.checkCanceled();
+                                        return Objects.requireNonNullElse(codeActions, Collections.emptyList());
+                                    });
+                        }));
     }
 
     /**
      * Create the LSP code action parameters for the given diagnostic and file.
      *
-     * @param diagnostic the diagnostic.
-     * @param file       the file.
+     * @param diagnostics the diagnostic.
+     * @param file        the file.
      * @return the LSP code action parameters for the given diagnostic and file.
      */
-    private static CodeActionParams createCodeActionParams(Diagnostic diagnostic, VirtualFile file) {
+    private static CodeActionParams createCodeActionParams(List<Diagnostic> diagnostics, VirtualFile file) {
         CodeActionParams params = new CodeActionParams();
         params.setTextDocument(LSPIJUtils.toTextDocumentIdentifier(file));
-        Range range = diagnostic.getRange();
+        // As diagnostic is sorted by the max range, the code action range parameter is the first diagnostic
+        Range range = diagnostics.get(0).getRange();

```

Comment:
either check the diagnostics list is not empty first, or add a comment explaining it's guaranteed  to be not empty

Reply:
fixed by adding comment, diagnostics list is never empty